kernel/src/utilities/arch_helpers.rs

```diff
@@ -0,0 +1,449 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
```

Comment:
Assuming we can move on #4229 quickly, I'd really like to separate that out and rebase this, because then it's much easier to see the difference between what the kernel is doing now and what this PR changes.